runtime/lua/vim/_system.lua

```diff
@@ -90,19 +90,24 @@ end
 
 local MAX_TIMEOUT = 2 ^ 31
 
---- @param timeout? integer
+--- @param timeout? integer timeout (in milliseconds), must be >= 0.
 --- @return vim.SystemCompleted
 function SystemObj:wait(timeout)
   local state = self._state
 
-  local done = vim.wait(timeout or state.timeout or MAX_TIMEOUT, function()
+  -- If timeout == 0, vim.wait() will return immediately and uv timer won't schedule on_exit()
+  -- in which state.result will be set
+  timeout = math.max(1, timeout or state.timeout or MAX_TIMEOUT)
+
+  local done = vim.wait(timeout, function()
     return state.result ~= nil
   end, nil, true)
 
   if not done then
-    -- Send sigkill since this cannot be caught
-    self:_timeout(SIG.KILL)
-    vim.wait(timeout or state.timeout or MAX_TIMEOUT, function()
+    -- Send SIGTERM since this cannot be caught
+    self:_timeout(SIG.TERM)
```

Comment:
SIGTERM *can* be caught, so the comment needs an update?

Reply:
Ah yes, this is why it was SIGKILL initially.

Reply:
Updated the comment. It'd be even better if we can do additional check if the process won't terminate after SIGTERM and send SIGKILL, but handling this seems a bit complicated so we could address this in the future. This PR more would aim to make the behavior of `vim.system(cmd, { timeout = ... }):wait()` and `vim.system(...):wait(timeout)` consistent, as well as making it work well with `timeout = 0`.